src/network/networkd-radv.c

```diff
@@ -443,20 +443,29 @@ static int radv_get_ip6dns(Network *network, struct in6_addr **dns,
 
 static int radv_set_dns(Link *link, Link *uplink) {
         _cleanup_free_ struct in6_addr *dns = NULL;
-        size_t n_dns;
         usec_t lifetime_usec;
+        size_t n_dns;
         int r;
 
         if (!link->network->router_emit_dns)
                 return 0;
 
         if (link->network->router_dns) {
-                dns = newdup(struct in6_addr, link->network->router_dns,
-                             link->network->n_router_dns);
+                struct in6_addr *p;
+
+                dns = new(struct in6_addr, link->network->n_router_dns);
                 if (!dns)
                         return -ENOMEM;
 
-                n_dns = link->network->n_router_dns;
+                p = dns;
+                for (size_t i = 0; i < link->network->n_router_dns; i++)
+                        if (in_addr_is_null(AF_INET6, (const union in_addr_union *) &link->network->router_dns[i])) {
+                                if (!in_addr_is_null(AF_INET6, (const union in_addr_union*) &link->ipv6ll_address))
+                                        *p++ = link->ipv6ll_address;
+                        } else
+                                *p++ = link->network->router_dns[i];
```

Comment:
can we write this as `*(p++)`. I mean, operator precedence is all great, but it's not obvious and it's easy to make this obvious, hence let's do it.